File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head){
        ListNode curr = head, temp = head;
        ListNode prev = null;

        while(curr != null){
            temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(head == null || head.next == null || left == right)return head;

        int count = 1;
        ListNode left_node = head, left_behind = null;
        ListNode curr = head;

        while(count<left){
            left_behind = curr;
            left_node = curr.next;
            curr = curr.next;
            count ++;
        }
        ListNode right_node = curr, right_after = null;
        
        while(count<=right){
            right_node = curr;
            right_after = curr.next;
            curr = curr.next;
            count++;
        }

        right_node.next = null;
        ListNode h = reverse(left_node);
        
        left_node.next = right_after;

        if(left_behind == null){
          return h;  
        }
        left_behind.next = h;

        return head; 

    }
}

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverse(ListNode *h){
        ListNode *p=NULL,*c=h, *n;
        
        while(c){
            n=c->next;
            c->next=p;
            p=c;
            c=n;
        }
        
        return p;
    }
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if(head==NULL || head->next==NULL || left==right)return head;
        int count=1;
        ListNode *l=head,*r,*p=NULL,*t,*n;
        while(count<left){
            p=l;
            l=l->next;
            count++;
        }
        r=l;
        while(count<right){
            r=r->next;
            count++;
        }
        n=r->next;
        r->next=NULL;
        ListNode *left1=reverse(l);
        
        if(p==NULL)head=left1;
        else
            p->next=left1;
        l->next=n;
        return head;
    }
};